src/video.c

```diff
@@ -129,8 +146,18 @@ void video_get_viewport_size(float *width, float *height) {
 
 void video_get_viewport(FloatRect *vp) {
 	video_get_viewport_size(&vp->w, &vp->h);
-	vp->x = (int)((video.current.width  - vp->w) / 2);
-	vp->y = (int)((video.current.height - vp->h) / 2);
+	if(video_highdpi_mode()) {
+		vp->x = (int)(video.current.width  - vp->w);
+		vp->y = (int)(video.current.height - vp->h);
+		vp->w = (int)(vp->w * 2);
+		vp->h = (int)(vp->h * 2);
+	} else {
```

Comment:
Now, another thing: **if I'm right about all of the above**, then maybe we actually do **not** want to use `SDL_WINDOW_ALLOW_HIGHDPI`, ever. It would make sense to simply tell the operating system to not automatically scale the window (in macOS's case, via the `Info.plist` entry as you have already done), and then have the SDL screen-space functions like `SDL_GetWindowSize` and `SDL_CreateWindow` operate in real pixel space, without magical implicit scaling. A benefit of this is that the displayed resolution in the options menu will actually be correct, and this should work on more platforms (SDL doesn't seem to have magic DPI scaling for Windows and X11 at least).

`SDL_GL_GetDrawableSize` would also coincide with `SDL_GetWindowSize` once again, although I think it's still more technically correct to express the viewport calculation in terms of the framebuffer size explicitly.

A downside of this approach is that we'll need to do a bit of extra work to make sure that the default window is not tiny. This could be done by pruning tiny windowed modes based on `SDL_GetDisplayDPI`, using about 100dpi as the baseline (roughly corresponds to a common 1920x1080 22" monitor). Or we might as well just default to fullscreen. Either way, that's a somewhat separate issue and can be dealt with later.

WebGL is still a mystery, whether `SDL_WINDOW_ALLOW_HIGHDPI` is used or not.